### ISSUE TITLE
FAIL: testExecuteTimeoutWithNastyChildren

### ISSUE THREAD

**Issue by opoplawski** [fail2ban/fail2ban#1155]
~~~~~
testExecuteTimeoutWithNastyChildren (fail2ban.tests.actiontestcase.CommandActionTest) ...  2015-08-10 14:15:58,450 140008252741376 Set action Test timeout = 60
 2015-08-10 14:15:58,451 140008252741376 Set actionstart =
 2015-08-10 14:15:58,451 140008252741376 Set actionban =
 2015-08-10 14:15:58,451 140008252741376 Set actionunban =
 2015-08-10 14:15:58,451 140008252741376 Set actioncheck =
 2015-08-10 14:15:58,451 140008252741376 Set actionstop =
 2015-08-10 14:15:58,451 140008252741376 Created <class 'fail2ban.server.action.CommandAction'>
FAIL
 2015-08-10 14:15:58,656 140008252741376
 2015-08-10 14:15:58,656 140008252741376 Nothing to do
testReplaceTag (fail2ban.tests.actiontestcase.CommandActionTest) ...  2015-08-10 14:15:58,657 140008252741376 Set action Test timeout = 60
 2015-08-10 14:15:58,657 140008252741376 Set actionstart =
 2015-08-10 14:15:58,657 140008252741376 Set actionban =
 2015-08-10 14:15:58,657 140008252741376 Set actionunban =
 2015-08-10 14:15:58,657 140008252741376 Set actioncheck =
 2015-08-10 14:15:58,657 140008252741376 Set actionstop =
 2015-08-10 14:15:58,657 140008252741376 Created <class 'fail2ban.server.action.CommandAction'>
 2015-08-10 14:15:58,657 140008252741376
 2015-08-10 14:15:58,657 140008252741376 Nothing to do
ok

======================================================================
FAIL: testExecuteTimeoutWithNastyChildren (fail2ban.tests.actiontestcase.CommandActionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./fail2ban/tests/actiontestcase.py", line 226, in testExecuteTimeoutWithNastyChildren
    RuntimeError, CommandAction.executeCmd, 'bash %s' % tmpFilename, timeout=.1)
AssertionError: RuntimeError not raised

----------------------------------------------------------------------
Ran 14 tests in 4.091s

FAILED (failures=1)
E: Failed with  /usr/bin/python2.7
~~~~~

Just starting to dig into this...

**Comment by sebres** [comment]
Hi,
I think, can be a timeout, possible to short (on busy or slow system).
I had twice another error with timeout, executed 1 second longer as expected (have already fixed it, for example here - https://github.com/sebres/fail2ban/commit/6ecd7ddddf1b9425cfac466a015f0c5f4255f1a6)

**Comment by yarikoptic** [comment]
nah - that one is a different issue... Here timeout is short but the point is that the script itself should sleep VERRY LONG (10000 sec), so whatever timeout time is specified, it should be shorter, thus that process must be killed and thus exception raised.  I only see a possibility that command does execute without sleeping for some reason, and completes without being killed.  in mycase -- this test didn't fail even when rebuilt across all debian/ubuntus -- so something more fun happening here

**Comment by sebres** [comment]
> and completes without being killed

I mean the timeout is so short (bash so slow), that the bash does not yet creates pipes or something other goes wrong (not in while cycle), and variable `retcode` has wrong value, because first will set here - https://github.com/fail2ban/fail2ban/blob/cb101e9f4abb4a7886ebd917f735c688ea758a00/fail2ban/server/action.py#L572

**Comment by opoplawski** [comment]
I think I understand what is happening, and it seems expected.  So I have no idea why the test wasn't failing elsewhere.  In the bash script you are trapping the TERM signal:
~~~~~
            f.write("""#!/bin/bash
        trap : HUP EXIT TERM

        echo "$$" > %s.pid
        echo "my pid $$ . sleeping lo-o-o-ong"
        sleep 10000
        """ % tmpFilename)
~~~~~
So when you kill the process group, only the sleep process dies directly due to SIGTERM.  The bash process continues and ends up exiting with a return code of 143:
~~~~~
32009 nanosleep({10000, 0},  <unfinished ...>
32009 <... nanosleep resumed> NULL)     = ? ERESTART_RESTARTBLOCK (Interrupted by signal)
32008 <... wait4 resumed> 0x7ffe3c52f8f0, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
32009 --- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=31982, si_uid=1744} ---
32008 --- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=31982, si_uid=1744} ---
32008 rt_sigreturn({mask=[CHLD]})       = -1 EINTR (Interrupted system call)
32008 wait4(-1,  <unfinished ...>
32009 +++ killed by SIGTERM +++
32008 <... wait4 resumed> [{WIFSIGNALED(s) && WTERMSIG(s) == SIGTERM}], 0, NULL) = 32009
....
32008 write(2, "Terminated\n", 11)      = 11
32008 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
32008 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=32009, si_uid=1744, si_status=SIGTERM, si_utime=0, si_stime=0} ---
32008 wait4(-1, 0x7ffe3c52f310, WNOHANG, NULL) = -1 ECHILD (No child processes)
...
32008 exit_group(143)                   = ?
32008 +++ exited with 143 +++
~~~~~
From bash's man page:
    When a command terminates on a fatal signal N, bash uses the value of 128+N as the exit status.
So bash is reporting that sleep exited with signal 15.  But poll() in action.py:executeCmd() then returns 143, and so no RuntimeError is generated.

So, why isn't this test failing on debian/ubuntu?  And I really don't know what type of situation is desired to be handled here.

**Comment by sebres** [comment]
Well, I think the `executeCmd` should always `return False` in case of error (retcode != 0), and this artificial test should be rewritten to `assertFalse` instead of `assertRaises`, because another real call of it always checks return code (https://github.com/fail2ban/fail2ban/search?utf8=%E2%9C%93&q=executeCmd). So throwing an error is really unexpected situation.

**Comment by opoplawski** [comment]
So, if I change it to assertFalse(), I then get:
~~~~~
FAIL: testExecuteTimeoutWithNastyChildren (fail2ban.tests.actiontestcase.CommandActionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./fail2ban/tests/actiontestcase.py", line 230, in testExecuteTimeoutWithNastyChildren
    self.assertTrue(self._is_logged('killed with SIGTERM'))
AssertionError: False is not true
~~~~~
Which brings up something that made debugging this harder than it might have been.  None of the logSys() messages from executeCmd appear to get logged.

**Comment by yarikoptic** [comment]
I would be damned -- issue replicates for me on Debian as well now ;) not sure how it all passed by before when I have built/tested all across, oh well

**Comment by sebres** [comment]
@yarikoptic I can not reproduce it on debian's ... :) where you had it? which debian (or python)?

**Comment by yarikoptic** [comment]
I believe it just happened for me straight on my laptop (mix of Debian testing/sid).  Now, upon initial run I also got a new one:
```
$> bin/fail2ban-testcases testExecuteTimeoutWithNastyChildren
Fail2ban 0.9.3.dev test suite. Python 2.7.10 (default, Jul  1 2015, 10:54:53) [GCC 4.9.2]. Please wait...
...........E..
======================================================================
ERROR: testExecuteTimeoutWithNastyChildren (fail2ban.tests.actiontestcase.CommandActionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./fail2ban/tests/actiontestcase.py", line 228, in testExecuteTimeoutWithNastyChildren
    self.assertFalse(pid_exists(getnastypid()))  # process should have been killed
  File "./fail2ban/tests/actiontestcase.py", line 221, in getnastypid
    with open(tmpFilename + '.pid') as f:
IOError: [Errno 2] No such file or directory: '/home/yoh/.tmp/fail2ban_Prh1iC.sh.pid'

----------------------------------------------------------------------
Ran 14 tests in 4.046s

FAILED (errors=1)
```
but then tested on another box, running jessie (with neurodebian etc), there that good old issue reproduces consistently
```
$> bin/fail2ban-testcases testExecuteTimeoutWithNastyChildren
Fail2ban 0.9.3.dev test suite. Python 2.7.9 (default, Mar  1 2015, 12:57:24) [GCC 4.9.2]. Please wait...
...........F..
======================================================================
FAIL: testExecuteTimeoutWithNastyChildren (fail2ban.tests.actiontestcase.CommandActionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./fail2ban/tests/actiontestcase.py", line 226, in testExecuteTimeoutWithNastyChildren
    RuntimeError, CommandAction.executeCmd, 'bash %s' % tmpFilename, timeout=.1)
AssertionError: RuntimeError not raised

----------------------------------------------------------------------
Ran 14 tests in 4.088s

FAILED (failures=1)
```

**Comment by opoplawski** [comment]
Any progress here?  I'm currently waiting for this to get fixed before upgrading to 0.9.3 in Fedora.

**Comment by yarikoptic** [comment]
On Sat, 12 Sep 2015, Orion Poplawski wrote:
> Any progress here? I'm currently waiting for this to get fixed before upgrading
> to 0.9.3 in Fedora.

I will look into it over the weekend some time
sorry about the delay ;)

**Comment by yarikoptic** [comment]
ok -- the main reason is that bash and dash reacted differently. bash returned 143 while dash with -15.  Logic in the code differentiates between negative (raises) and positive error code exists (returns True). Different systems of mine had different /bin/sh symlink thus on some it didn't fail.  I think that @sebres is right and we should simplify logic there to not raise but rather consistently return False -- inspecting history and program logic I didn't find where catching exceptions higher in the stack would be beneficial really.  I will send a PR later today/tomorrow

**Comment by yarikoptic** [comment]
@opoplawski for 0.9.3 the best would be just to "bandage" the test:  https://github.com/yarikoptic/fail2ban/commit/3187de29336b4a01e4dcd20f98253e1db913ee7a .  Best would be not to change behavior of fail2ban I think.  I will later prep PR with "proper" fix... now family time

**Comment by opoplawski** [comment]
Will do.  Thanks.

**Comment by sebres** [comment]
This has been continued: as I had rebased my branch https://github.com/fail2ban/fail2ban/tree/f2b-perfom-prepare-716 relative current muster (where this bug should be fixed) I got this bug again (I think because it works faster resp. sleeps fewer). No exception but false assert:
```
self.assertFalse(pid_exists(getnastypid()))  # process should have been killed
```
In addition I had sporadically another one - because the bash does not succeed to create a pid-file inside bash-script and got timeouted.
```
File "./fail2ban/tests/actiontestcase.py", line 221, in getnastypid
    with open(tmpFilename + '.pid') as f:
IOError: [Errno 2] No such file or directory
```

Both are repaired and fixed in my current version of this branch (previously #1116).

ATM no idea how I can fix (resp. reproduce it) it in current master.
